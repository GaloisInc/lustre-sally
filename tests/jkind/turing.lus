-- Turing machine simulated in Lustre.
-- 4-state busy beaver on a 2-symbol language:
--
-- http://en.wikipedia.org/wiki/Busy_beaver
--
-- Expected output:
--   INVALID PROPERTY: cex || K = 108
--
--
-- The encoding of the tape uses three integer variables: left, head,
-- and right.
--
--   * The symbol underneath the tape head is stored in 'head'.
--
--   * Everything to the left of the tape head is stored in 'left'
--     using a special base-10 (for readability) encoding. The least
--     significant digit of 'left' is the symbol one cell to the left
--     of the tape head, the second least significant digit of 'left'
--     is the symbol two cells to the left of the tape head, etc.

--   * Everything to the right of the tape head is stored in 'right'
--     using a similar base-10 notation. The least significant digit
--     of 'right' is the symbol one cell to the right of the tape
--     head, the second least significant digit of 'right' is the
--     symbol two cells to the right of the tape head, etc.
--
-- For example, consider the tape:
--
--   ... 0 1 0 1 1 [0] 0 1 1 0 1 0 ...
--
-- Where the '...' parts are all zero and the tape head is at [0].
-- Then left = 1011, head = 0, and right = 10110.
--
-- Since the tape only ever has a finite number of non-zero symbols on
-- it, this encoding is sufficient. Moreover, we easily shift the tape
-- left and right using 'div 10' and 'mod 10'.

type state = enum { Halt, A, B, C, D };
type move = enum { L, R };
type symbol = int;

node main() returns (left, right, head : int);
var
  write : symbol;
  move : move;
  state : state;

  cex : bool;
let
  (left, head, right) = (0, 0, 0) ->
    if pre state = Halt then pre (left, head, right) else
    if move = L then (pre left div 10, pre left mod 10, 10*(pre right) + write)
                else (10*(pre left) + write, pre right mod 10, pre right div 10);

  (write, move, state) = (0, L, A) -> -- Only the initial state matters here
         if pre (state, head) = (A, 0) then (1, R, B)
    else if pre (state, head) = (A, 1) then (1, L, B)
    else if pre (state, head) = (B, 0) then (1, L, A)
    else if pre (state, head) = (B, 1) then (0, L, C)
    else if pre (state, head) = (C, 0) then (1, R, Halt)
    else if pre (state, head) = (C, 1) then (1, L, D)
    else if pre (state, head) = (D, 0) then (1, R, D)
    else if pre (state, head) = (D, 1) then (0, R, A)
    else (0, L, Halt); -- Only the Halt state matters here

  cex = state <> Halt;
  --%PROPERTY cex;
tel;
