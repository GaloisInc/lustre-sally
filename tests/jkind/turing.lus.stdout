Core Lustre
===========

node ()
  returns (left, right, head)
  assumes ()
  shows (cex)
let
  pre_state$26 : int
    = pre state
   
  pre_head$24 : int
    = pre head
   
  pre_state$25 : int
    = pre state
   
  pre_head$23 : int
    = pre head
   
  pre_state$24 : int
    = pre state
   
  pre_head$22 : int
    = pre head
   
  pre_state$23 : int
    = pre state
   
  pre_head$21 : int
    = pre head
   
  pre_state$22 : int
    = pre state
   
  pre_head$20 : int
    = pre head
   
  pre_state$21 : int
    = pre state
   
  pre_head$19 : int
    = pre head
   
  pre_state$20 : int
    = pre state
   
  pre_head$18 : int
    = pre head
   
  pre_state$19 : int
    = pre state
   
  pre_head$17 : int
    = pre head
   
  write : int
    = 0 -> ITE(And(Eq(pre_state$26, 1), Eq(pre_head$24, 0)), 1, ITE(And(Eq(pre_state$25, 1), Eq(pre_head$23, 1)), 1, ITE(And(Eq(pre_state$24, 2), Eq(pre_head$22, 0)), 1, ITE(And(Eq(pre_state$23, 2), Eq(pre_head$21, 1)), 0, ITE(And(Eq(pre_state$22, 3), Eq(pre_head$20, 0)), 1, ITE(And(Eq(pre_state$21, 3), Eq(pre_head$19, 1)), 1, ITE(And(Eq(pre_state$20, 4), Eq(pre_head$18, 0)), 1, ITE(And(Eq(pre_state$19, 4), Eq(pre_head$17, 1)), 0, 0))))))))
   
  pre_state$18 : int
    = pre state
   
  pre_head$16 : int
    = pre head
   
  pre_state$17 : int
    = pre state
   
  pre_head$15 : int
    = pre head
   
  pre_state$16 : int
    = pre state
   
  pre_head$14 : int
    = pre head
   
  pre_state$15 : int
    = pre state
   
  pre_head$13 : int
    = pre head
   
  pre_state$14 : int
    = pre state
   
  pre_head$12 : int
    = pre head
   
  pre_state$13 : int
    = pre state
   
  pre_head$11 : int
    = pre head
   
  pre_state$12 : int
    = pre state
   
  pre_head$10 : int
    = pre head
   
  pre_state$11 : int
    = pre state
   
  pre_head$9 : int
    = pre head
   
  move : int
    = 0 -> ITE(And(Eq(pre_state$18, 1), Eq(pre_head$16, 0)), 1, ITE(And(Eq(pre_state$17, 1), Eq(pre_head$15, 1)), 0, ITE(And(Eq(pre_state$16, 2), Eq(pre_head$14, 0)), 0, ITE(And(Eq(pre_state$15, 2), Eq(pre_head$13, 1)), 0, ITE(And(Eq(pre_state$14, 3), Eq(pre_head$12, 0)), 1, ITE(And(Eq(pre_state$13, 3), Eq(pre_head$11, 1)), 0, ITE(And(Eq(pre_state$12, 4), Eq(pre_head$10, 0)), 1, ITE(And(Eq(pre_state$11, 4), Eq(pre_head$9, 1)), 1, 0))))))))
   
  pre_state$10 : int
    = pre state
   
  pre_left$3 : int
    = pre left
   
  pre_left$2 : int
    = pre left
   
  pre_left$1 : int
    = pre left
   
  left : int
    = 0 -> ITE(Eq(pre_state$10, 0), pre_left$3, ITE(Eq(move, 0), Div(pre_left$2, 10), Add(Mul(10, pre_left$1), write)))
   
  pre_state$9 : int
    = pre state
   
  pre_right$3 : int
    = pre right
   
  pre_right$2 : int
    = pre right
   
  pre_right$1 : int
    = pre right
   
  right : int
    = 0 -> ITE(Eq(pre_state$9, 0), pre_right$3, ITE(Eq(move, 0), Add(Mul(10, pre_right$2), write), Div(pre_right$1, 10)))
   
  pre_state$8 : int
    = pre state
   
  pre_head$8 : int
    = pre head
   
  pre_left : int
    = pre left
   
  pre_right : int
    = pre right
   
  head : int
    = 0 -> ITE(Eq(pre_state$8, 0), pre_head$8, ITE(Eq(move, 0), Mod(pre_left, 10), Mod(pre_right, 10)))
   
  pre_state$7 : int
    = pre state
   
  pre_head$7 : int
    = pre head
   
  pre_state$6 : int
    = pre state
   
  pre_head$6 : int
    = pre head
   
  pre_state$5 : int
    = pre state
   
  pre_head$5 : int
    = pre head
   
  pre_state$4 : int
    = pre state
   
  pre_head$4 : int
    = pre head
   
  pre_state$3 : int
    = pre state
   
  pre_head$3 : int
    = pre head
   
  pre_state$2 : int
    = pre state
   
  pre_head$2 : int
    = pre head
   
  pre_state$1 : int
    = pre state
   
  pre_head$1 : int
    = pre head
   
  pre_state : int
    = pre state
   
  pre_head : int
    = pre head
   
  state : int
    = 1 -> ITE(And(Eq(pre_state$7, 1), Eq(pre_head$7, 0)), 2, ITE(And(Eq(pre_state$6, 1), Eq(pre_head$6, 1)), 2, ITE(And(Eq(pre_state$5, 2), Eq(pre_head$5, 0)), 1, ITE(And(Eq(pre_state$4, 2), Eq(pre_head$4, 1)), 3, ITE(And(Eq(pre_state$3, 3), Eq(pre_head$3, 0)), 0, ITE(And(Eq(pre_state$2, 3), Eq(pre_head$2, 1)), 4, ITE(And(Eq(pre_state$1, 4), Eq(pre_head$1, 0)), 4, ITE(And(Eq(pre_state, 4), Eq(pre_head, 1)), 1, 0))))))))
   
  cex : bool
    = Neq(state, 0)
tel
Sally Model
===========

(define-state-type
   S
   (
     (cex Bool)
     (head Int)
     (left Int)
     (move Int)
     (right Int)
     (state Int)
     (write Int)
     (|gal-initializing| Bool)
     (|head:init| Bool)
     (|left:init| Bool)
     (|move:init| Bool)
     (|pre_head:10| Int)
     (|pre_head:11| Int)
     (|pre_head:12| Int)
     (|pre_head:13| Int)
     (|pre_head:14| Int)
     (|pre_head:15| Int)
     (|pre_head:16| Int)
     (|pre_head:17| Int)
     (|pre_head:18| Int)
     (|pre_head:19| Int)
     (|pre_head:1| Int)
     (|pre_head:20| Int)
     (|pre_head:21| Int)
     (|pre_head:22| Int)
     (|pre_head:23| Int)
     (|pre_head:24| Int)
     (|pre_head:2| Int)
     (|pre_head:3| Int)
     (|pre_head:4| Int)
     (|pre_head:5| Int)
     (|pre_head:6| Int)
     (|pre_head:7| Int)
     (|pre_head:8| Int)
     (|pre_head:9| Int)
     (|pre_head| Int)
     (|pre_left:1| Int)
     (|pre_left:2| Int)
     (|pre_left:3| Int)
     (|pre_left| Int)
     (|pre_right:1| Int)
     (|pre_right:2| Int)
     (|pre_right:3| Int)
     (|pre_right| Int)
     (|pre_state:10| Int)
     (|pre_state:11| Int)
     (|pre_state:12| Int)
     (|pre_state:13| Int)
     (|pre_state:14| Int)
     (|pre_state:15| Int)
     (|pre_state:16| Int)
     (|pre_state:17| Int)
     (|pre_state:18| Int)
     (|pre_state:19| Int)
     (|pre_state:1| Int)
     (|pre_state:20| Int)
     (|pre_state:21| Int)
     (|pre_state:22| Int)
     (|pre_state:23| Int)
     (|pre_state:24| Int)
     (|pre_state:25| Int)
     (|pre_state:26| Int)
     (|pre_state:2| Int)
     (|pre_state:3| Int)
     (|pre_state:4| Int)
     (|pre_state:5| Int)
     (|pre_state:6| Int)
     (|pre_state:7| Int)
     (|pre_state:8| Int)
     (|pre_state:9| Int)
     (|pre_state| Int)
     (|right:init| Bool)
     (|state:init| Bool)
     (|write:init| Bool))
   ())
(define-transition-system
   TS
   S
   (and
      (= |gal-initializing| true)
      (= |write:init| true)
      (= |move:init| true)
      (= |left:init| true)
      (= |right:init| true)
      (= |head:init| true)
      (= |state:init| true))
   (and
      (= next.|gal-initializing| false)
      (= next.|pre_state:26| state.state)
      (= next.|pre_head:24| state.head)
      (= next.|pre_state:25| state.state)
      (= next.|pre_head:23| state.head)
      (= next.|pre_state:24| state.state)
      (= next.|pre_head:22| state.head)
      (= next.|pre_state:23| state.state)
      (= next.|pre_head:21| state.head)
      (= next.|pre_state:22| state.state)
      (= next.|pre_head:20| state.head)
      (= next.|pre_state:21| state.state)
      (= next.|pre_head:19| state.head)
      (= next.|pre_state:20| state.state)
      (= next.|pre_head:18| state.head)
      (= next.|pre_state:19| state.state)
      (= next.|pre_head:17| state.head)
      (ite
         state.|write:init|
         (= next.write 0)
         (=
            next.write
            (ite
               (and
                  (= next.|pre_state:26| 1)
                  (= next.|pre_head:24| 0))
               1
               (ite
                  (and
                     (= next.|pre_state:25| 1)
                     (= next.|pre_head:23| 1))
                  1
                  (ite
                     (and
                        (= next.|pre_state:24| 2)
                        (= next.|pre_head:22| 0))
                     1
                     (ite
                        (and
                           (= next.|pre_state:23| 2)
                           (= next.|pre_head:21| 1))
                        0
                        (ite
                           (and
                              (= next.|pre_state:22| 3)
                              (= next.|pre_head:20| 0))
                           1
                           (ite
                              (and
                                 (= next.|pre_state:21| 3)
                                 (= next.|pre_head:19| 1))
                              1
                              (ite
                                 (and
                                    (= next.|pre_state:20| 4)
                                    (= next.|pre_head:18| 0))
                                 1
                                 (ite
                                    (and
                                       (= next.|pre_state:19| 4)
                                       (= next.|pre_head:17| 1))
                                    0
                                    0))))))))))
      (= next.|write:init| false)
      (= next.|pre_state:18| state.state)
      (= next.|pre_head:16| state.head)
      (= next.|pre_state:17| state.state)
      (= next.|pre_head:15| state.head)
      (= next.|pre_state:16| state.state)
      (= next.|pre_head:14| state.head)
      (= next.|pre_state:15| state.state)
      (= next.|pre_head:13| state.head)
      (= next.|pre_state:14| state.state)
      (= next.|pre_head:12| state.head)
      (= next.|pre_state:13| state.state)
      (= next.|pre_head:11| state.head)
      (= next.|pre_state:12| state.state)
      (= next.|pre_head:10| state.head)
      (= next.|pre_state:11| state.state)
      (= next.|pre_head:9| state.head)
      (ite
         state.|move:init|
         (= next.move 0)
         (=
            next.move
            (ite
               (and
                  (= next.|pre_state:18| 1)
                  (= next.|pre_head:16| 0))
               1
               (ite
                  (and
                     (= next.|pre_state:17| 1)
                     (= next.|pre_head:15| 1))
                  0
                  (ite
                     (and
                        (= next.|pre_state:16| 2)
                        (= next.|pre_head:14| 0))
                     0
                     (ite
                        (and
                           (= next.|pre_state:15| 2)
                           (= next.|pre_head:13| 1))
                        0
                        (ite
                           (and
                              (= next.|pre_state:14| 3)
                              (= next.|pre_head:12| 0))
                           1
                           (ite
                              (and
                                 (= next.|pre_state:13| 3)
                                 (= next.|pre_head:11| 1))
                              0
                              (ite
                                 (and
                                    (= next.|pre_state:12| 4)
                                    (= next.|pre_head:10| 0))
                                 1
                                 (ite
                                    (and
                                       (= next.|pre_state:11| 4)
                                       (= next.|pre_head:9| 1))
                                    1
                                    0))))))))))
      (= next.|move:init| false)
      (= next.|pre_state:10| state.state)
      (= next.|pre_left:3| state.left)
      (= next.|pre_left:2| state.left)
      (= next.|pre_left:1| state.left)
      (ite
         state.|left:init|
         (= next.left 0)
         (=
            next.left
            (ite
               (= next.|pre_state:10| 0)
               next.|pre_left:3|
               (ite
                  (= next.move 0)
                  (ite
                     (< 10 0)
                     (-
                        (to_int
                           (/
                              (to_real next.|pre_left:2|)
                              (to_real
                                 (- 10)))))
                     (to_int
                        (/
                           (to_real next.|pre_left:2|)
                           (to_real 10))))
                  (+
                     (* 10 next.|pre_left:1|)
                     next.write)))))
      (= next.|left:init| false)
      (= next.|pre_state:9| state.state)
      (= next.|pre_right:3| state.right)
      (= next.|pre_right:2| state.right)
      (= next.|pre_right:1| state.right)
      (ite
         state.|right:init|
         (= next.right 0)
         (=
            next.right
            (ite
               (= next.|pre_state:9| 0)
               next.|pre_right:3|
               (ite
                  (= next.move 0)
                  (+
                     (* 10 next.|pre_right:2|)
                     next.write)
                  (ite
                     (< 10 0)
                     (-
                        (to_int
                           (/
                              (to_real next.|pre_right:1|)
                              (to_real
                                 (- 10)))))
                     (to_int
                        (/
                           (to_real next.|pre_right:1|)
                           (to_real 10))))))))
      (= next.|right:init| false)
      (= next.|pre_state:8| state.state)
      (= next.|pre_head:8| state.head)
      (= next.|pre_left| state.left)
      (= next.|pre_right| state.right)
      (ite
         state.|head:init|
         (= next.head 0)
         (=
            next.head
            (ite
               (= next.|pre_state:8| 0)
               next.|pre_head:8|
               (ite
                  (= next.move 0)
                  (mod next.|pre_left| 10)
                  (mod next.|pre_right| 10)))))
      (= next.|head:init| false)
      (= next.|pre_state:7| state.state)
      (= next.|pre_head:7| state.head)
      (= next.|pre_state:6| state.state)
      (= next.|pre_head:6| state.head)
      (= next.|pre_state:5| state.state)
      (= next.|pre_head:5| state.head)
      (= next.|pre_state:4| state.state)
      (= next.|pre_head:4| state.head)
      (= next.|pre_state:3| state.state)
      (= next.|pre_head:3| state.head)
      (= next.|pre_state:2| state.state)
      (= next.|pre_head:2| state.head)
      (= next.|pre_state:1| state.state)
      (= next.|pre_head:1| state.head)
      (= next.|pre_state| state.state)
      (= next.|pre_head| state.head)
      (ite
         state.|state:init|
         (= next.state 1)
         (=
            next.state
            (ite
               (and
                  (= next.|pre_state:7| 1)
                  (= next.|pre_head:7| 0))
               2
               (ite
                  (and
                     (= next.|pre_state:6| 1)
                     (= next.|pre_head:6| 1))
                  2
                  (ite
                     (and
                        (= next.|pre_state:5| 2)
                        (= next.|pre_head:5| 0))
                     1
                     (ite
                        (and
                           (= next.|pre_state:4| 2)
                           (= next.|pre_head:4| 1))
                        3
                        (ite
                           (and
                              (= next.|pre_state:3| 3)
                              (= next.|pre_head:3| 0))
                           0
                           (ite
                              (and
                                 (= next.|pre_state:2| 3)
                                 (= next.|pre_head:2| 1))
                              4
                              (ite
                                 (and
                                    (= next.|pre_state:1| 4)
                                    (= next.|pre_head:1| 0))
                                 4
                                 (ite
                                    (and
                                       (= next.|pre_state| 4)
                                       (= next.|pre_head| 1))
                                    1
                                    0))))))))))
      (= next.|state:init| false)
      (=
         next.cex
         (not
            (= next.state 0)))))

(query
   TS
   (or |gal-initializing| cex))

[Lustre] Validating properties:
[Lustre] Property cex... [Error] Sally error:
/tmp/sallyXXX22970-0.mcmt(292)  : error 3 : , at offset 18
    near [Index: 1081 (Start: 1147405191-Stop: 1147405191) ='(', type<14> Line: 292 LinePos:18]
     : cannot match to any predicted input...
Parse error: /tmp/sallyXXX22970-0.mcmt:292:18: Parse error: /tmp/sallyXXX22970-0.mcmt:292:18: Parse error.

[Error] ExitFailure 1

